diff --git a/libbb/lineedit.c b/libbb/lineedit.c
index c87d606..ec7d4d6 100644
--- a/libbb/lineedit.c
+++ b/libbb/lineedit.c
@@ -1659,10 +1659,17 @@ static void remember_in_history(char *str)
 		return;
 	if (str[0] == '\0')
 		return;
-	i = state->cnt_history;
+
 	/* Don't save dupes */
-	if (i && strcmp(state->history[i-1], str) == 0)
-		return;
+	for (i = 0; i < state->cnt_history; i++)
+		if (strcmp(state->history[i], str) == 0) {
+			int j;
+			unsigned int str = state->history[i];
+			for (j = i + 1; j < state->cnt_history; j++)
+				state->history[j - 1] = state->history[j];
+			state->history[j] = str;
+			return;
+		}

 	free(state->history[state->max_history]); /* redundant, paranoia */
 	state->history[state->max_history] = NULL; /* redundant, paranoia */
